File: output/xCodeEval_remove_8000/C++/Java/2ba8bd30f8f86980a15291e0ca997648.java
import java.io.*;
import java.util.*;

class DSum {
    static class Solver {
        private final BufferedReader in;
        private final PrintWriter out;

        public Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        private void solve() throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            int n = Integer.parseInt(st.nextToken());
            int k = Integer.parseInt(st.nextToken());

            long[] sums = new long[n];
            List<List<Integer>> arrs = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                st = new StringTokenizer(in.readLine());
                int t = Integer.parseInt(st.nextToken());
                List<Integer> arr = new ArrayList<>();
                for (int j = 0; j < t; j++) {
                    arr.add(Integer.parseInt(st.nextToken()));
                }
                if (t > k) {
                    arr = arr.subList(0, k);
                }
                sums[i] = arr.stream().mapToLong(Integer::longValue).sum();
                arrs.add(arr);
            }

            long[][] dps = new long[13][k + 1];
            for (int i = 0; i < 13; i++) {
                Arrays.fill(dps[i], 0);
            }

            long ans = rec(0, n, k, dps, sums, arrs);
            out.println(ans);
        }

        private long rec(int l, int r, int k, long[][] dps, long[] sums, List<List<Integer>> arrs) {
            long ans = 0;
            List<Integer> carr = arrs.get(l);
            long[] dp = dps[0];
            if (l + 1 == r) {
                ans = dp[k];
                long sum = 0;
                for (int i = 0; i < Math.min(k, carr.size()); i++) {
                    sum += carr.get(i);
                    ans = Math.max(ans, dp[k - i - 1] + sum);
                }
            } else {
                int m = (l + r) / 2;
                doCopy(dps, l, m, k);
                ans = Math.max(ans, rec(m, r, k, dps, sums, arrs));
                doCopy(dps, m, r, k);
                ans = Math.max(ans, rec(l, m, k, dps, sums, arrs));
            }
            return ans;
        }

        private void doCopy(long[][] dps, int l, int r, int k) {
            for (int i = l; i < r; i++) {
                addToknap(i, dps, k);
            }
        }

        private void addToknap(int i, long[][] dps, int k) {
            long[] dp = dps[i + 1];
            long v = sums[i];
            List<Integer> carr = arrs.get(i);
            int k1 = carr.size();
            for (int j = k; j >= k1; j--) {
                dp[j] = Math.max(dp[j], dp[j - k1] + v);
            }
        }
    }

    public static void main(String[] args) {
        try {
            BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
            PrintWriter out = new PrintWriter(System.out);
            Solver solver = new Solver(in, out);
            solver.solve();
            out.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/2ba8bd30f8f86980a15291e0ca997648.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define max(a, b) ((a) > (b) ? (a) : (b))

typedef long long ll;
typedef struct {
    int size;
    int* arr;
} vi;
typedef struct {
    int size;
    vi* arr;
} vvi;
typedef struct {
    int size;
    ll* arr;
} vll;
typedef struct {
    int size;
    vll* arr;
} vvll;

int k;
vvll dps;
vll sums;
vvi arrs;

void do_copy(int dpi) {
    memcpy(dps.arr[dpi + 1].arr, dps.arr[dpi].arr, sizeof(ll) * (k + 1));
}

void add_to_knap(int i, int dpi) {
    ll v = sums.arr[i];
    int k1 = arrs.arr[i].size;
    for (int j = k; j >= k1; j--) {
        dps.arr[dpi].arr[j] = max(dps.arr[dpi].arr[j], dps.arr[dpi].arr[j - k1] + v);
    }
}

ll rec(int l, int r, int dpi) {
    ll ans = 0;
    vi carr = arrs.arr[l];
    vll dp = dps.arr[dpi];
    if (l + 1 == r) {
        ans = max(ans, dp.arr[k]);
        ll sum = 0;
        for (int i = 0; i < min(k, carr.size); i++) {
            sum += carr.arr[i];
            ans = max(ans, dp.arr[k - i - 1] + sum);
        }
    } else {
        int m = (l + r) / 2;
        do_copy(dpi);
        for (int i = l; i < m; i++) {
            add_to_knap(i, dpi + 1);
        }
        ans = max(ans, rec(m, r, dpi + 1));
        do_copy(dpi);
        for (int i = m; i < r; i++) {
            add_to_knap(i, dpi + 1);
        }
        ans = max(ans, rec(l, m, dpi + 1));
    }
    return ans;
}

void solve() {
    int n;
    scanf("%d %d", &n, &k);
    sums.size = n;
    sums.arr = (ll*)malloc(sizeof(ll) * n);
    arrs.size = n;
    arrs.arr = (vi*)malloc(sizeof(vi) * n);
    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        arrs.arr[i].size = t;
        arrs.arr[i].arr = (int*)malloc(sizeof(int) * t);
        for (int j = 0; j < t; j++) {
            scanf("%d", &arrs.arr[i].arr[j]);
        }
        if (t > k) {
            arrs.arr[i].size = k;
        }
        sums.arr[i] = 0;
        for (int j = 0; j < arrs.arr[i].size; j++) {
            sums.arr[i] += arrs.arr[i].arr[j];
        }
    }
    dps.size = 13;
    dps.arr = (vll*)malloc(sizeof(vll) * 13);
    for (int i = 0; i < 13; i++) {
        dps.arr[i].size = k + 1;
        dps.arr[i].arr = (ll*)malloc(sizeof(ll) * (k + 1));
    }
    printf("%lld\n", rec(0, n, 0));
}

int main() {
    solve();
    return 0;
}
